From f4a74939db385f69e897074cdd9c5a1e3129518e Mon Sep 17 00:00:00 2001
From: Leo Yu-Chi Liang <ycliang@andestech.com>
Date: Wed, 14 Aug 2024 14:08:52 +0800
Subject: [PATCH 160/172] platform: generic: andes: Add ae350 system suspend
 device template (#116)

Add system suspend device to support AE350 SUSP sbi extension.

"hart_resume" hook needs to be implemented for main hart resume procedure.

Main hart:
suspend: Kernel SUSP SBI ecall --> SUSP: sbi_system_suspend --> ae350 suspend hook
resume: _start_warm --> init_warmboot --> init_warm_resume --> ae350 hart resume hook
	--> restore registers setting --> ...

Other harts:
suspend: Kernel HSM SBI ecall --> HSM: hart_stop --> ae350 hart stop
resume: enable cache --> _start_warm --> init_warmboot --> init_warm_startup
	--> ae350 early init hook --> restore registers setting --> ...

Signed-off-by: Leo Yu-Chi Liang <ycliang@andestech.com>
---
 platform/generic/andes/ae350.c | 38 +++++++++++++++++++++++++++-------
 1 file changed, 31 insertions(+), 7 deletions(-)

diff --git a/platform/generic/andes/ae350.c b/platform/generic/andes/ae350.c
index 5d8cb54c29e6..e13296965544 100644
--- a/platform/generic/andes/ae350.c
+++ b/platform/generic/andes/ae350.c
@@ -19,6 +19,7 @@
 #include <sbi/sbi_hsm.h>
 #include <sbi/sbi_ipi.h>
 #include <sbi/sbi_init.h>
+#include <sbi/sbi_system.h>
 #include <andes/andes_sbi.h>
 #include <andes/andes.h>
 #include <andes/andes_pma.h>
@@ -111,13 +112,35 @@ fail:
 	return 0;
 }
 
-static const struct sbi_hsm_device andes_smu = {
-	.name	      = "andes_smu",
-	.hart_start   = ae350_hart_start,
-	.hart_stop    = ae350_hart_stop,
+static void ae350_hart_resume(void)
+{
+	return;
+}
+
+static const struct sbi_hsm_device andes_smu_hsm = {
+	.name		= "andes_smu",
+	.hart_start	= ae350_hart_start,
+	.hart_stop	= ae350_hart_stop,
+	.hart_resume	= ae350_hart_resume,
+};
+
+static int ae350_system_suspend_check(u32 sleep_type)
+{
+	return 0;
+}
+
+static int ae350_system_suspend(u32 sleep_type, unsigned long mmode_resume_addr)
+{
+	return 0;
+}
+
+static struct sbi_system_suspend_device andes_smu_susp = {
+	.name			= "andes_smu",
+	.system_suspend_check	= ae350_system_suspend_check,
+	.system_suspend		= ae350_system_suspend,
 };
 
-static void ae350_hsm_device_init(void)
+static void ae350_smu_device_init(void)
 {
 	int rc;
 	void *fdt;
@@ -128,7 +151,8 @@ static void ae350_hsm_device_init(void)
 				   "andestech,atcsmu");
 
 	if (!rc) {
-		sbi_hsm_set_device(&andes_smu);
+		sbi_hsm_set_device(&andes_smu_hsm);
+		sbi_system_suspend_set_device(&andes_smu_susp);
 	}
 }
 
@@ -138,7 +162,7 @@ static int ae350_final_init(bool cold_boot, const struct fdt_match *match)
 		return 0;
 
 	pma_init();
-	ae350_hsm_device_init();
+	ae350_smu_device_init();
 	trigger_init();
 
 	return 0;
-- 
2.34.1

